cedar-policy-core/src/evaluator/err.rs

```diff
@@ -16,140 +16,216 @@
 
 use crate::ast::*;
 use crate::parser::Loc;
-use itertools::Itertools;
-use miette::{Diagnostic, LabeledSpan};
+use miette::Diagnostic;
 use nonempty::{nonempty, NonEmpty};
 use smol_str::SmolStr;
 use std::sync::Arc;
 use thiserror::Error;
 
-/// An error generated while evaluating an expression
-#[derive(Debug, PartialEq, Eq, Clone, Error)]
-#[error("{error_kind}")]
-pub struct EvaluationError {
-    /// The kind of error that occurred
-    error_kind: EvaluationErrorKind,
```

Comment:
745 doesn't directly prescribe getting rid of `*Kind`, but it does indirectly, by recommending that our errors be enums.  I do think that getting rid of `EvaluationErrorKind` leads to a better experience for public-API consumers, who get to match on `EvaluationError` directly instead of having to match on `.error_kind()`.  The cost is that `source_loc` is duplicated in each error struct, but honestly I think that makes the code cleaner; the previous design had a couple instances where source loc would be specified in two places and assumed to align.

I still like the `*Kind` pattern for `Expr`.